### PR TITLE
[AMDGPU] Extend llvm.amdgcn.update.dpp intrinsic to support f64

### PR DESCRIPTION
**Pull request by vikramRH**
Follow up patch to https://github.com/llvm/llvm-project/pull/89217, before we make changes to atomic optimizer. 
@arsenm , @jayfoad , Do you see any concerns with changes here ?

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIInstructions.td [hunk 1]
```diff
@@ -3338,14 +3338,17 @@ def : GCNPat <
                         (as_i1timm $bound_ctrl))
 >;
 
-def : GCNPat <
-  (i64 (int_amdgcn_update_dpp i64:$old, i64:$src, timm:$dpp_ctrl, timm:$row_mask,
+class UpdateDPP64Pat<ValueType vt> : GCNPat <
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstructions.td:3341]: Nit: instead of creating a class you could do this with `foreach vt = [i64, f64] in ...`

Comment (vikramRH) [reply llvm/lib/Target/AMDGPU/SIInstructions.td]: Done

File: llvm/lib/Target/AMDGPU/SIInstructions.td [hunk 2]
```diff
@@ -3338,13 +3338,15 @@ def : GCNPat <
                         (as_i1timm $bound_ctrl))
 >;
 
+foreach vt = [i64, f64] in {
```
[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIInstructions.td:3341]: For the 16/32 cases, we currently have Reg32Types. Should create a new Reg64Types and start populating it for this

[the remark this answers is not in the document]
Comment (vikramRH) [reply llvm/lib/Target/AMDGPU/SIInstructions.td]: Done

**Comment by vikramRH** [comment]
Added new tests for v2i32, float2 and ptr

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.update.dpp.ll [hunk 3]
```diff
@@ -81,6 +81,45 @@ define amdgpu_kernel void @update_dppf64_test(ptr addrspace(1) %arg, double %in1
...
+; GCN-DAG: v_mov_b32_dpp v{{[0-9]+}}, v[[SRC_HI]] quad_perm:[1,0,0,0] row_mask:0x1 bank_mask:0x1{{$}}
+define amdgpu_kernel void @update_dpp_p0_test(ptr addrspace(1) %arg, ptr %in1, ptr %in2) {
+  %id = tail call i32 @llvm.amdgcn.workitem.id.x()
+  %gep = getelementptr inbounds ptr, ptr addrspace(1) %arg, i32 %id
+  %load = load ptr, ptr addrspace(1) %gep
+  %tmp0 = call ptr @llvm.amdgcn.update.dpp.p0(ptr %in1, ptr %load, i32 1, i32 1, i32 1, i1 false) #0
+  store ptr %tmp0, ptr addrspace(1) %gep
+  ret void
+}
+
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/llvm.amdgcn.update.dpp.ll:122]: Should have some 32-bit pointer tests too, don't think we have those

File: llvm/lib/Target/AMDGPU/SIRegisterInfo.td [hunk 4]
```diff
@@ -586,6 +586,7 @@ class RegisterTypes<list<ValueType> reg_types> {
 
 def Reg16Types : RegisterTypes<[i16, f16, bf16]>;
 def Reg32Types : RegisterTypes<[i32, f32, v2i16, v2f16, v2bf16, p2, p3, p5, p6]>;
+def Reg64Types : RegisterTypes<[i64, f64, v2i32, v2f32, p0]>;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIRegisterInfo.td:589]: We'll need to do something about other address spaces for globalisel, eventually

**Comment by arsenm** [comment]
As a follow up, can you make AMDGPUInstCombineIntrinsic fold update.dpp.i64 (bitcast (x to i64)) to the different typed intrinsic

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.update.dpp.ll (hunk 3 above)
Comment (vikramRH) [reply llvm/test/CodeGen/AMDGPU/llvm.amdgcn.update.dpp.ll:122]: Added new tests

**Comment by vikramRH** [comment]
> As a follow up, can you make AMDGPUInstCombineIntrinsic fold update.dpp.i64 (bitcast (x to i64)) to the different typed intrinsic

Will do, should this be a separate patch ?

**Comment by jayfoad** [comment]
LGTM, though I wish there was a better way to implement it then duplicating the selection pattern for every 64-bit type.

Is there any reason the sdag and globalisel tests can't be combined into a single file with two RUN lines?

**Comment by jayfoad** [comment]
> > As a follow up, can you make AMDGPUInstCombineIntrinsic fold update.dpp.i64 (bitcast (x to i64)) to the different typed intrinsic
> 
> Will do, should this be a separate patch ?

Yes.

**Comment by arsenm** [comment]
> LGTM, though I wish there was a better way to implement it then duplicating the selection pattern for every 64-bit type.

I've had an open issue for ages to add a type matcher that only cares about the size, but nobody ever gets anywhere with it

**Review by arsenm — APPROVED** [review]

**Pull request merged**